spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java

```diff
@@ -172,17 +173,29 @@ private String jobDesc() {
     Column nameEqual = actualFileName.equalTo(validFileName);
     Column actualContains = actualFileDF.col("file_path").contains(validFileDF.col("file_path"));
     Column joinCond = nameEqual.and(actualContains);
-    List<String> orphanFiles = actualFileDF.join(validFileDF, joinCond, "leftanti")
-        .as(Encoders.STRING())
-        .collectAsList();
-
-    Tasks.foreach(orphanFiles)
-        .noRetry()
-        .executeWith(deleteExecutorService)
-        .suppressFailureWhenFinished()
-        .onFailure((file, exc) -> LOG.warn("Failed to delete file: {}", file, exc))
-        .run(deleteFunc::accept);
+    Dataset<String> orphanDF = actualFileDF.join(validFileDF, joinCond, "leftanti")
+            .as(Encoders.STRING());
+    return deleteFiles(orphanDF);
+  }
 
+  private BaseDeleteOrphanFilesActionResult deleteFiles(Dataset<String> orphanDF) {
+    boolean streamResults = PropertyUtil.propertyAsBoolean(options(), STREAM_RESULTS, false);
+    Iterator<String> itr;
+    if (streamResults) {
+      itr = orphanDF.toLocalIterator();
+    } else {
+      itr = orphanDF.collectAsList().iterator();
+    }
+    List<String> orphanFiles = Lists.newArrayList();
```

Comment:
[question] elements in the list is being added concurrently do we need to ensure thread safety ?
```suggestion
    List<String> orphanFiles = Collections.synchronizedList(Lists.newArrayList());
```
refering to this [SO Answer](https://stackoverflow.com/questions/2715983/concurrent-threads-adding-to-arraylist-at-same-time-what-happens)